### ISSUE TITLE
Masked multi-head attention in Transformer code

### ISSUE THREAD

**Issue by khyun0630** [tensorflow/tensor2tensor#941]
I want to find the code of "masked" multi-head attention 
for "masking out (setting to −∞)" all values in the input of the softmax which correspond to illegal connections in decode part.

But, I cannot find the difference in (self) multi-head attentions of encode and decode parts.

**Comment by colmantse** [comment]
Hey, I think its at the prepare_decoder part, typically, transformer use causal_decoder_self_attention, which would invoke the attention_bias_lower_triangle and mask everything that is beyond the current id.

**Comment by colmantse** [comment]
@lukaszkaiser on a side note however, may i inquire if during inference stage, at every decoding step, decode_length of the target would always be 1 (i.e, using the last decoded token) instead of all previous decoded tokens, as specified in both symbol_to_logits_fn and preprocess_targets, or am I missing something? I checked beamsearch.beamsearch but it seems the symbol_to_logits_fn is still from the transformer.py so in the end, the target sequence that got to the bottom of the decoder stack during every inference step would still be the last decoded token basically.
